modules/text2vec-voyageai/vectorizer/objects.go

```diff
@@ -13,32 +13,65 @@ package vectorizer
 
 import (
 	"context"
-	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/tiktoken-go"
+	"github.com/weaviate/weaviate/modules/text2vec-openai/clients"
+	"github.com/weaviate/weaviate/usecases/modulecomponents/batch"
+	"github.com/weaviate/weaviate/usecases/modulecomponents/text2vecbase"
 
 	"github.com/weaviate/weaviate/entities/models"
 	"github.com/weaviate/weaviate/entities/moduletools"
 	"github.com/weaviate/weaviate/modules/text2vec-voyageai/ent"
 	objectsvectorizer "github.com/weaviate/weaviate/usecases/modulecomponents/vectorizer"
-	libvectorizer "github.com/weaviate/weaviate/usecases/vectorizer"
 )
 
-type Vectorizer struct {
-	client           Client
-	objectVectorizer *objectsvectorizer.ObjectVectorizer
-}
+const (
+	MaxObjectsPerBatch    = 128 // https://docs.voyageai.com/docs/embeddings#python-api
+	OpenAIMaxTimePerBatch = float64(10)
+)
+
+func New(client text2vecbase.BatchClient, logger logrus.FieldLogger) *text2vecbase.BatchVectorizer {
+	batchTokenizer := func(ctx context.Context, objects []*models.Object, skipObject []bool, cfg moduletools.ClassConfig, objectVectorizer *objectsvectorizer.ObjectVectorizer) ([]string, []int, bool, error) {
+		texts := make([]string, len(objects))
+		tokenCounts := make([]int, len(objects))
+		icheck := ent.NewClassSettings(cfg)
+
+		// the encoding is different than OpenAI, but the code is not available in Go and too complicated to port.
+		// However, it is close to the OpenAI encoding and with a small margin, we can use the OpenAI token count
+		// https://docs.voyageai.com/docs/tokenization
+		tke, err := tiktoken.EncodingForModel("text-embedding-ada-002")
```

Comment:
Did you consult this with VoyageAI team? that they use this tokenizer?

Reply:
From the bottom of the linked page:

> tiktoken is the open-source version of OpenAI's tokenizer. Voyage uses a different tokenizer, which is the same as [Llama 2](https://ai.meta.com/llama/). Therefore, our tokenizer may generate a different list of tokens for a given text compared to tiktoken. Statistically, the number of tokens produced by our tokenizer is on average 1.1 - 1.2 times that of tiktoken. To determine the exact number of tokens, please use the Client.count_tokens() function.

This is just used to not hit the token limit, so I am using this to be safe:
```
			tokenCounts[i] = int(float32(clients.GetTokensCount(icheck.Model(), text, tke)) * 1.3)
```